h/api_client/api_client.py

```diff
@@ -0,0 +1,67 @@
+# -*- coding: utf-8 -*-
+import requests
+
+
+class APIError(Exception):
+
+    """Raised when a request to the API fails for any reason."""
+
+    pass
+
+
+class ConnectionError(APIError):
+
+    """Raised when a request to the API fails because of a network problem."""
+
+    pass
+
+
+class Timeout(APIError):
+
+    """Raised when a request to the API times out."""
+
+    pass
+
+
+class Client(object):
+
+    """A client for the Hypothesis API.
+
+    Provides simple methods for calling the API over HTTP.
+
+    Note that this class is deliberately decoupled from the Hypothesis Pyramid
+    app and from Pyramid.
+
+    """
+
+    def __init__(self, base_url):
+        """Initialize a new API client.
+
+        :param base_url: The base URL to the Hypothesis API instance to use.
+            For example: ``"http://hypothesis.is/api"``.
+
+        """
+        self.base_url = base_url
+
+    def get(self, path, params=None):
```

Comment:
I kind of want to improve the abstraction here and not let GET and POST etc leak. So `client.search(limit=10)` instead of `client.get("/search", params={"limit": 10})`.

The problem with this is what we have to add a method to our API client for every API method, and a param for every param, and keep them up to date.

You can do a `__getattr__()` trick on `Client` so that `client.search(limit=10)` gets turned into `requests.get("/search", params={"limit": 10})` without actually defining a `.search()` method or params on `Client`. (It'll also let you do `client.foo(bar=10)` and it'll attempt to get `/foo` and get an error back from the API which it can turn into an exception to return to the caller.)

The problem I ran into with this was how can the client decide whether to do a GET or a POST?

A middle-ground might be to have a constant `Client.__POST_PATHS__ = ["/create", "/update"]` so that if it's going to request any of those paths it does a post, for everything else it does a get. We still have to maintain that list and add new paths to it whenever we add a new post method to the API, but it minimises the amount of work we have to do.

What do you think?